### DISCUSSION TITLE
How to import Monaco editor into Nuxt3 app?

### DISCUSSION THREAD

**Discussion by th3killwillkill** [nuxt/nuxt#17271] [Questions] [answered]
Hi,

I am having problem with [Monaco editor](https://github.com/microsoft/monaco-editor) implementation into Nuxt3 app.

i googled this topic, but there is not many resources about that, everything I tried always end up with this

nuxi dev:
```
[nuxt] [request error] navigator is not defined
  at $id_a8dc1548 (./.nuxt/dist/server/server.mjs:13990:19)  
  at async __instantiateModule__ (./.nuxt/dist/server/server.mjs:107163:3)
```

and http server output:
```
{
  "statusCode": 404,
  "statusMessage": "Not Found",
  "stack": []
}
```
I tried using Monaco Editor Webpack Loader Plugin (like [this](https://github.com/FE-Mars/monaco-editor-vue/issues/11)), but I am not sure if I even implemented that correctly.

(one weird thing, import seems successful if I change page code and let hot reload do its thing, but after full reload its same as output mentioned before)

**Could somebody show me a simplest implementation?** (Nuxt3 nuxi project initialized and successful import of Monaco editor)

Thank you

**Comment by Diizzayy** [comment, marked as answer]
> [nuxt] [request error] navigator is not defined
  at $id_a8dc1548 (./.nuxt/dist/server/server.mjs:13990:19)  
  at async __instantiateModule__ (./.nuxt/dist/server/server.mjs:107163:3)

@th3killwillkill You need to ensure that the initialization code only runs on client and after the page has completed loading, hence Vue's `onMounted` hook can be leveraged.

- Install Monaco Editor
```sh
yarn add monaco-editor @monaco-editor/loader
```

- Usage

```vue
<template>
  <main>
    <div id="editor" ref="editor" />
  </main>
</template>

<script lang="ts" setup>
const editor = ref()

onMounted(async () => {
  const loader = await import('@monaco-editor/loader').then(m => m?.default)

  const monaco = await loader.init()

  monaco.editor.create(editor.value, {
    theme: 'vs-dark',
    language: 'typescript',
    value: "const foo = 'bar'"
  })
})
</script>

<style>
#editor {
  height: 100vh;
}
</style>
```

**Reply by xsky13** [reply]
Is there a way to access the monaco editor outside of the `onMounted` function?

**Reply by Diizzayy** [reply]
@xsky13 Ofcourse, you just need to declare the monaco variable outside the `onMounted` scope, as only it's initialization needs to happen therein. After doing so, you'll be able to access the monaco instance elsewhere.

**Reply by xsky13** [reply]
Something like this?
```
let monacoEditor = ref();
onMounted(async () => {
  const loader = await import('@monaco-editor/loader').then(m => m?.default)

  const monaco = await loader.init()

  monacoEditor.value = monaco.editor.create(editor.value, {
    theme: 'vs-dark',
    language: 'typescript',
    value: "const foo = 'bar'"
  })
})

// Use monacoEditor outside of onMounted
console.log(monacoEditor.value)
```

**Reply by xsky13** [reply]
@Diizzayy  btw, that 👆 didn't work

**Reply by Diizzayy** [reply]
@xsky13 The console log would essentially trigger before the instance is set by `onMounted`, here is a better example below.

```vue
<script lang="ts" setup>
import type { Monaco } from '@monaco-editor/loader'

const editor = ref()
const monaco = ref<Monaco>()

onMounted(async () => {
  const loader = await import('@monaco-editor/loader').then(m => m?.default)

  monaco.value = await loader.init()

  monaco.value.editor.create(editor.value, {
    theme: 'vs-dark',
    language: 'typescript',
    value: "const foo = 'bar'"
  })
})

function setTheme() {
  monaco.value.editor.defineTheme('myCustomTheme', {
    base: 'vs', // can also be vs-dark or hc-black
    inherit: true, // can also be false to completely replace the builtin rules
    rules: [
      { token: 'comment', foreground: 'ffa500', fontStyle: 'italic underline' },
      { token: 'comment.js', foreground: '008800', fontStyle: 'bold' },
      { token: 'comment.css', foreground: '0000ff' } // will inherit fontStyle from `comment` above
    ],
    colors: {
      'editor.background': '#000000',
      'editor.foreground': '#ffffff',
    }
  })

  monaco.value.editor.setTheme('myCustomTheme')
}
</script>
```

**Reply by xsky13** [reply]
Ok, thank you very much

**Reply by xsky13** [reply]
Sorry for bothering again, but I'm trying to make a function to change the editor code by clicking a button, but it doesn't work. What am I doing wrong?
```
const changeContent = (content) => {
    monaco.value.editor.getModel().setValue(content);
}
```
The `setTheme()` function works perfectly well, but the `changeContent` function gives me an error:
```
TypeError: undefined is not an object (evaluating 'R.toString')
```

**Reply by xsky13** [reply]
Fixed it by by using `getModals()[0]`
